nixpkgs_review/report.py

```diff
@@ -55,84 +55,134 @@ def ensure(self) -> Path:
+                            path,
+                        ],
+                        stdout=f,
+                    )
+                    if nix_log.returncode == 0:
+                        break
+
+
+def _serialize_attrs(attrs: list[Attr]) -> list[str]:
+    return list(map(lambda a: a.name, attrs))
+
+
+class SystemReport:
+    def __init__(self, attrs: list[Attr]) -> None:
+        self.broken: list[Attr] = []
+        self.failed: list[Attr] = []
+        self.non_existent: list[Attr] = []
+        self.blacklisted: list[Attr] = []
+        self.tests: list[Attr] = []
+        self.built: list[Attr] = []
+
+        for attr in attrs:
+            if attr.broken:
+                self.broken.append(attr)
+            elif attr.blacklisted:
+                self.blacklisted.append(attr)
+            elif not attr.exists:
+                self.non_existent.append(attr)
+            elif attr.name.startswith("nixosTests."):
+                self.tests.append(attr)
+            elif not attr.was_build():
+                self.failed.append(attr)
             else:
-                symlink_source = failed_results.ensure().joinpath(attr.name)
-            if os.path.lexists(symlink_source):
-                symlink_source.unlink()
-            symlink_source.symlink_to(attr.path)
+                self.built.append(attr)
 
-        for path in [f"{attr.drv_path}^*", attr.path]:
-            if not path:
-                continue
-            with open(
-                logs.ensure().joinpath(attr.name + ".log"), "w+", encoding="utf-8"
-            ) as f:
-                nix_log = subprocess.run(
-                    [
-                        "nix",
-                        "--extra-experimental-features",
-                        "nix-command",
-                        "log",
-                        path,
-                    ],
-                    stdout=f,
-                )
-                if nix_log.returncode == 0:
-                    break
+    def serialize(self) -> dict[str, list[str]]:
+        return {
+            "broken": _serialize_attrs(self.broken),
+            "non-existent": _serialize_attrs(self.non_existent),
+            "blacklisted": _serialize_attrs(self.blacklisted),
+            "failed": _serialize_attrs(self.failed),
+            "built": _serialize_attrs(self.built),
+            "tests": _serialize_attrs(self.tests),
+        }
+
+
+def order_reports(reports: dict[System, SystemReport]) -> dict[System, SystemReport]:
+    """Ensure that systems are always ordered consistently in reports"""
+    system_order: dict[System, int] = {
+        "x86_64-linux": 0,
+        "aarch64-linux": 1,
+        "x86_64-darwin": 2,
+        "aarch64-darwin": 3,
+    }
+
+    def sort_func(system: System) -> str:
+        arch, os = system.split("-")
+        return f"{os}-{arch}"
+
+    return dict(
+        sorted(
+            reports.items(),
+            key=lambda item: "".join(reversed(item[0])),
```

Comment:
```suggestion
            key=lambda item: reversed(item[0].split("-")),
```

Reply:
^ This should not sort architectures more naturally.

Reply:
Ok, adding `reverse=True` I was able to achieve the desired order without needing a static table.